In [1]:
import pandas as pd
import seaborn as sns

In [3]:
df = sns.load_dataset('titanic')

In [5]:
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

### 컬럼 추가

In [6]:
df_copy = df.copy()

In [7]:
df_copy['VIP'] = False

In [9]:
df_copy.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone    VIP  
0    man        True  NaN  Southampton    no  False  False  
1  woman       False    C    Cherbourg   yes  False  False  
2  woman       False  NaN  Southampton   yes   True  False  
3  woman       False    C  Southampton   yes  False  False  
4    man        True  NaN  Southampton    no   True  False

In [10]:
# family 컬럼 만들기
df_copy['family'] = df_copy['sibsp'] + df_copy['parch']
df_copy.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone    VIP  family  
0    man        True  NaN  Southampton    no  False  False       1  
1  woman       False    C    Cherbourg   yes  False  False       1  
2  woman       False  NaN  Southampton   yes   True  False       0  
3  woman       False    C  Southampton   yes  False  False       1  
4    man        True  NaN  Southampton    no   True  False       0

In [11]:
# 글자의 이어붙이기
df_copy['gender'] = df_copy['sex'] + '-' + df_copy['who']
df_copy.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone    VIP  family  \
0    man        True  NaN  Southampton    no  False  False       1   
1  woman       False    C    Cherbourg   yes  False  False       1   
2  woman       False  NaN  Southampton   yes   True  False       0   
3  woman       False    C  Southampton   yes  False  False       1   
4    man        True  NaN  Southampton    no   True  False       0   

         gender  
0      male-man  
1  female-woman  
2  female-woman  
3  female-woman  
4      male-man

In [15]:
df_copy['f/a'] = round(df_copy['fare'] / df_copy['age'], 2) # round: 소수의 자릿수 계산, 반올림 처리
df_copy.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone    VIP  family  \
0    man        True  NaN  Southampton    no  False  False       1   
1  woman       False    C    Cherbourg   yes  False  False       1   
2  woman       False  NaN  Southampton   yes   True  False       0   
3  woman       False    C  Southampton   yes  False  False       1   
4    man        True  NaN  Southampton    no   True  False       0   

         gender   f/a  
0      male-man  0.33  
1  female-woman  1.88  
2  female-woman  0.30  
3  female-woman  1.52  
4      male-man  0.23

### 컬럼 삭제

### 행 삭제

In [16]:
df_copy.drop(1) # 인덱스 지정하면 됨

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
5           0       3    male   NaN      0      0   8.4583        Q   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone    VIP  family  \
0      man        True  NaN  Southampton    no  False  False       1   
2    woman       False  NaN  Southampton   yes   True  False       0   
3    woman       False    C  Southampton   yes  False  False       1   
4      man        True  NaN  Southampton    no   True  False       0   
5      man        True  NaN   Queenstown    no   True  False       0   
..     ...         ...  ...          ...   ...    ...    ...     ...   
886    man        True  NaN  Southampton    no   True  False       0   
887  woman       False    B  Southampton   yes   True  False       0   
888  woman       False  NaN  Southampton    no  False  False       3   
889    man        True    C    Cherbourg   yes   True  False       0   
890    man        True  NaN   Queenstown    no   True  False       0   

           gender   f/a  
0        male-man  0.33  
2    female-woman  0.30  
3    female-woman  1.52  
4        male-man  0.23  
5        male-man   NaN  
..            ...   ...  
886      male-man  0.48  
887  female-woman  1.58  
888  female-woman   NaN  
889      male-man  1.15  
890      male-man  0.24  

[890 rows x 19 columns]

In [18]:
df_copy.head() # 원본은 바뀌지 않음

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone    VIP  family  \
0    man        True  NaN  Southampton    no  False  False       1   
1  woman       False    C    Cherbourg   yes  False  False       1   
2  woman       False  NaN  Southampton   yes   True  False       0   
3  woman       False    C  Southampton   yes  False  False       1   
4    man        True  NaN  Southampton    no   True  False       0   

         gender   f/a  
0      male-man  0.33  
1  female-woman  1.88  
2  female-woman  0.30  
3  female-woman  1.52  
4      male-man  0.23

In [19]:
# 범위 지정해서 삭제도 가능

In [20]:
df_copy.drop(range(5))

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
5           0       3    male   NaN      0      0   8.4583        Q   Third   
6           0       1    male  54.0      0      0  51.8625        S   First   
7           0       3    male   2.0      3      1  21.0750        S   Third   
8           1       3  female  27.0      0      2  11.1333        S   Third   
9           1       2  female  14.0      1      0  30.0708        C  Second   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone    VIP  family  \
5      man        True  NaN   Queenstown    no   True  False       0   
6      man        True    E  Southampton    no   True  False       0   
7    child       False  NaN  Southampton    no  False  False       4   
8    woman       False  NaN  Southampton   yes  False  False       2   
9    child       False  NaN    Cherbourg   yes  False  False       1   
..     ...         ...  ...          ...   ...    ...    ...     ...   
886    man        True  NaN  Southampton    no   True  False       0   
887  woman       False    B  Southampton   yes   True  False       0   
888  woman       False  NaN  Southampton    no  False  False       3   
889    man        True    C    Cherbourg   yes   True  False       0   
890    man        True  NaN   Queenstown    no   True  False       0   

           gender    f/a  
5        male-man    NaN  
6        male-man   0.96  
7      male-child  10.54  
8    female-woman   0.41  
9    female-child   2.15  
..            ...    ...  
886      male-man   0.48  
887  female-woman   1.58  
888  female-woman    NaN  
889      male-man   1.15  
890      male-man   0.24  

[886 rows x 19 columns]

In [21]:
df_copy.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone    VIP  family  \
0    man        True  NaN  Southampton    no  False  False       1   
1  woman       False    C    Cherbourg   yes  False  False       1   
2  woman       False  NaN  Southampton   yes   True  False       0   
3  woman       False    C  Southampton   yes  False  False       1   
4    man        True  NaN  Southampton    no   True  False       0   

         gender   f/a  
0      male-man  0.33  
1  female-woman  1.88  
2  female-woman  0.30  
3  female-woman  1.52  
4      male-man  0.23

In [22]:
# 리스트 지정으로 삭제도 가능

In [23]:
df_copy.drop([1, 3, 5, 7, 9])

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
2           1       3  female  26.0      0      0   7.9250        S   Third   
4           0       3    male  35.0      0      0   8.0500        S   Third   
6           0       1    male  54.0      0      0  51.8625        S   First   
8           1       3  female  27.0      0      2  11.1333        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone    VIP  family  \
0      man        True  NaN  Southampton    no  False  False       1   
2    woman       False  NaN  Southampton   yes   True  False       0   
4      man        True  NaN  Southampton    no   True  False       0   
6      man        True    E  Southampton    no   True  False       0   
8    woman       False  NaN  Southampton   yes  False  False       2   
..     ...         ...  ...          ...   ...    ...    ...     ...   
886    man        True  NaN  Southampton    no   True  False       0   
887  woman       False    B  Southampton   yes   True  False       0   
888  woman       False  NaN  Southampton    no  False  False       3   
889    man        True    C    Cherbourg   yes   True  False       0   
890    man        True  NaN   Queenstown    no   True  False       0   

           gender   f/a  
0        male-man  0.33  
2    female-woman  0.30  
4        male-man  0.23  
6        male-man  0.96  
8    female-woman  0.41  
..            ...   ...  
886      male-man  0.48  
887  female-woman  1.58  
888  female-woman   NaN  
889      male-man  1.15  
890      male-man  0.24  

[886 rows x 19 columns]

### 열 삭제

In [24]:
df_copy.drop('VIP', axis=1)

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  family        gender  \
0      man        True  NaN  Southampton    no  False       1      male-man   
1    woman       False    C    Cherbourg   yes  False       1  female-woman   
2    woman       False  NaN  Southampton   yes   True       0  female-woman   
3    woman       False    C  Southampton   yes  False       1  female-woman   
4      man        True  NaN  Southampton    no   True       0      male-man   
..     ...         ...  ...          ...   ...    ...     ...           ...   
886    man        True  NaN  Southampton    no   True       0      male-man   
887  woman       False    B  Southampton   yes   True       0  female-woman   
888  woman       False  NaN  Southampton    no  False       3  female-woman   
889    man        True    C    Cherbourg   yes   True       0      male-man   
890    man        True  NaN   Queenstown    no   True       0      male-man   

      f/a  
0    0.33  
1    1.88  
2    0.30  
3    1.52  
4    0.23  
..    ...  
886  0.48  
887  1.58  
888   NaN  
889  1.15  
890  0.24  

[891 rows x 18 columns]

In [25]:
df_copy.drop(['deck', 'VIP', 'alive'], axis=1).head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male  embark_town  alone  family        gender   f/a  
0    man        True  Southampton  False       1      male-man  0.33  
1  woman       False    Cherbourg  False       1  female-woman  1.88  
2  woman       False  Southampton   True       0  female-woman  0.30  
3  woman       False  Southampton  False       1  female-woman  1.52  
4    man        True  Southampton   True       0      male-man  0.23

In [26]:
df_copy.drop(['deck', 'alive'], axis=1, inplace=True)
df_copy.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male  embark_town  alone    VIP  family        gender   f/a  
0    man        True  Southampton  False  False       1      male-man  0.33  
1  woman       False    Cherbourg  False  False       1  female-woman  1.88  
2  woman       False  Southampton   True  False       0  female-woman  0.30  
3  woman       False  Southampton  False  False       1  female-woman  1.52  
4    man        True  Southampton   True  False       0      male-man  0.23

### 데이터 타입

In [27]:
df_copy = df.copy()

In [28]:
df_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


In [29]:
df_copy['who'].value_counts()

who
man      537
woman    271
child     83
Name: count, dtype: int64

In [38]:
df_copy['who'] = df_copy['who'].astype('category') 
# astype() 강제적으로 형변환할 때 사용하는 함수
df_copy['who'].dtype
# 원본이 바뀌지는 않음

CategoricalDtype(categories=['child', 'man', 'woman'], ordered=False, categories_dtype=object)

In [40]:
# pandas 코드가 바라보는 모습이 카테고리의 형태임
# child, man, woman -> 1, 2, 3 이런 식으로...
df_copy['who'].cat.codes

0      1
1      2
2      2
3      2
4      1
      ..
886    1
887    2
888    2
889    1
890    1
Length: 891, dtype: int8

In [41]:
# 카테고리 이름 수정 가능
df_copy['who'] = df_copy['who'].cat.rename_categories(['아이', '남자', '여자'])
df_copy['who'].value_counts()

who
남자    537
여자    271
아이     83
Name: count, dtype: int64

### datetime

pd.date_range() # 날짜 데이터를 여러 개 만들 때 사용함

- pd.date_range('20250101', periods=891) # 숫자 직접 넣지 말기!
- pd.date_range('20250101', periods=df.shape[0])

In [47]:
dates = pd.date_range('20250101', periods=df.shape[0], freq='15h')
dates

DatetimeIndex(['2025-01-01 00:00:00', '2025-01-01 15:00:00',
               '2025-01-02 06:00:00', '2025-01-02 21:00:00',
               '2025-01-03 12:00:00', '2025-01-04 03:00:00',
               '2025-01-04 18:00:00', '2025-01-05 09:00:00',
               '2025-01-06 00:00:00', '2025-01-06 15:00:00',
               ...
               '2026-07-05 15:00:00', '2026-07-06 06:00:00',
               '2026-07-06 21:00:00', '2026-07-07 12:00:00',
               '2026-07-08 03:00:00', '2026-07-08 18:00:00',
               '2026-07-09 09:00:00', '2026-07-10 00:00:00',
               '2026-07-10 15:00:00', '2026-07-11 06:00:00'],
              dtype='datetime64[ns]', length=891, freq='15h')

In [48]:
df_copy['date'] = dates
df_copy.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class who  \
0         0       3    male  22.0      1      0   7.2500        S  Third  남자   
1         1       1  female  38.0      1      0  71.2833        C  First  여자   
2         1       3  female  26.0      0      0   7.9250        S  Third  여자   
3         1       1  female  35.0      1      0  53.1000        S  First  여자   
4         0       3    male  35.0      0      0   8.0500        S  Third  남자   

   adult_male deck  embark_town alive  alone                date  
0        True  NaN  Southampton    no  False 2025-01-01 00:00:00  
1       False    C    Cherbourg   yes  False 2025-01-01 15:00:00  
2       False  NaN  Southampton   yes   True 2025-01-02 06:00:00  
3       False    C  Southampton   yes  False 2025-01-02 21:00:00  
4        True  NaN  Southampton    no   True 2025-01-03 12:00:00

In [49]:
df_copy['date'].dt.year

0      2025
1      2025
2      2025
3      2025
4      2025
       ... 
886    2026
887    2026
888    2026
889    2026
890    2026
Name: date, Length: 891, dtype: int32

In [50]:
df_copy['date'].dt.dayofweek

0      2
1      2
2      3
3      3
4      4
      ..
886    2
887    3
888    4
889    4
890    5
Name: date, Length: 891, dtype: int32

In [51]:
# 일요일 데이터를 모두 가져오고 싶다면?
cond = df_copy['date'].dt.dayofweek == 6
df_copy.loc[cond] # 필터링

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
7           0       3    male   2.0      3      1  21.0750        S   Third   
18          0       3  female  31.0      1      0  18.0000        S   Third   
19          1       3  female   NaN      0      0   7.2250        C   Third   
29          0       3    male   NaN      0      0   7.8958        S   Third   
30          0       1    male  40.0      0      0  27.7208        C   First   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
859         0       3    male   NaN      0      0   7.2292        C   Third   
869         1       3    male   4.0      1      1  11.1333        S   Third   
870         0       3    male  26.0      0      0   7.8958        S   Third   
880         1       2  female  25.0      0      1  26.0000        S  Second   
881         0       3    male  33.0      0      0   7.8958        S   Third   

    who  adult_male deck  embark_town alive  alone                date  
7    아이       False  NaN  Southampton    no  False 2025-01-05 09:00:00  
18   여자       False  NaN  Southampton    no  False 2025-01-12 06:00:00  
19   여자       False  NaN    Cherbourg   yes   True 2025-01-12 21:00:00  
29   남자        True  NaN  Southampton    no   True 2025-01-19 03:00:00  
30   남자        True  NaN    Cherbourg    no   True 2025-01-19 18:00:00  
..   ..         ...  ...          ...   ...    ...                 ...  
859  남자        True  NaN    Cherbourg    no   True 2026-06-21 21:00:00  
869  아이       False  NaN  Southampton   yes  False 2026-06-28 03:00:00  
870  남자        True  NaN  Southampton    no   True 2026-06-28 18:00:00  
880  여자       False  NaN  Southampton   yes  False 2026-07-05 00:00:00  
881  남자        True  NaN  Southampton    no   True 2026-07-05 15:00:00  

[127 rows x 16 columns]

## 시계열 정보 접근 연습하기 - 자전거 데이터

In [52]:
df = pd.read_csv('data/seoul_bicycle.csv')
df.head()

대여일자  대여소번호      대여소명  대여구분코드  성별    연령대코드  이용건수      운동량    탄소량  \
0  Jan-20-2020      3      중랑센터  일일(회원)   M  AGE_003     3    61.82   0.52   
1  Jan-20-2020      3      중랑센터  일일(회원)   M  AGE_004     1    39.62   0.28   
2  Jan-20-2020      3      중랑센터      정기   M  AGE_005     3   430.85   4.01   
3  Jan-20-2020      5  상암센터 정비실  일일(회원)  \N  AGE_005     2     1.79   0.02   
4  Jan-20-2020      5  상암센터 정비실      정기   F  AGE_003     1  4501.96  45.47   

       이동거리  이용시간  
0    2230.0    75  
1    1220.0    15  
2   17270.0    53  
3      90.0    33  
4  196010.0    64

In [53]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 327231 entries, 0 to 327230
Data columns (total 11 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   대여일자    327231 non-null  object 
 1   대여소번호   327231 non-null  int64  
 2   대여소명    327231 non-null  object 
 3   대여구분코드  327231 non-null  object 
 4   성별      272841 non-null  object 
 5   연령대코드   327231 non-null  object 
 6   이용건수    327231 non-null  int64  
 7   운동량     327231 non-null  object 
 8   탄소량     327231 non-null  object 
 9   이동거리    327231 non-null  float64
 10  이용시간    327231 non-null  int64  
dtypes: float64(1), int64(3), object(7)
memory usage: 27.5+ MB


In [59]:
df_copy = df.copy()

In [61]:
df_copy['대여일자'] = pd.to_datetime(df_copy['대여일자'], format='%b-%d-%Y')
df_copy['대여일자'].info()
# 내가 원하는 정보를 시계열로 변환하는 함수

<class 'pandas.core.series.Series'>
RangeIndex: 327231 entries, 0 to 327230
Series name: 대여일자
Non-Null Count   Dtype         
--------------   -----         
327231 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 2.5 MB


In [62]:
cond = df_copy['대여일자'].dt.dayofweek == 2
df_copy.loc[cond]

대여일자  대여소번호       대여소명  대여구분코드  성별    연령대코드  이용건수      운동량  \
249610 2020-05-20      3       중랑센터      정기   M  AGE_005     6  1566.25   
249611 2020-05-20      5   상암센터 정비실  일일(회원)  \N  AGE_005     7     3.78   
249612 2020-05-20      5   상암센터 정비실  일일(회원)   M  AGE_004    14   305.14   
249613 2020-05-20      5   상암센터 정비실  일일(회원)   M  AGE_005     2        0   
249614 2020-05-20      5   상암센터 정비실      정기  \N  AGE_008     1        0   
...           ...    ...        ...     ...  ..      ...   ...      ...   
327226 2020-05-20   9997  9997.강남센터  일일(회원)   M  AGE_006     7   689.57   
327227 2020-05-20   9997  9997.강남센터      정기  \N  AGE_002     1        0   
327228 2020-05-20   9997  9997.강남센터      정기  \N  AGE_004     2    19.96   
327229 2020-05-20   9997  9997.강남센터      정기   M  AGE_003     6    43.77   
327230 2020-05-20   9997  9997.강남센터      정기   M  AGE_004     6  4735.63   

          탄소량      이동거리  이용시간  
249610  14.56   62780.0   107  
249611   0.03     190.0    96  
249612    2.8   12040.0   185  
249613      0       0.0    14  
249614      0       0.0     0  
...       ...       ...   ...  
327226   6.22   26790.0   206  
327227      0       0.0    25  
327228   0.16     700.0    10  
327229   0.34    1470.0   134  
327230  32.64  140690.0   283  

[77621 rows x 11 columns]

In [63]:
df_copy.head()

대여일자  대여소번호      대여소명  대여구분코드  성별    연령대코드  이용건수      운동량    탄소량  \
0 2020-01-20      3      중랑센터  일일(회원)   M  AGE_003     3    61.82   0.52   
1 2020-01-20      3      중랑센터  일일(회원)   M  AGE_004     1    39.62   0.28   
2 2020-01-20      3      중랑센터      정기   M  AGE_005     3   430.85   4.01   
3 2020-01-20      5  상암센터 정비실  일일(회원)  \N  AGE_005     2     1.79   0.02   
4 2020-01-20      5  상암센터 정비실      정기   F  AGE_003     1  4501.96  45.47   

       이동거리  이용시간  
0    2230.0    75  
1    1220.0    15  
2   17270.0    53  
3      90.0    33  
4  196010.0    64

In [64]:
df_copy['운동량'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 327231 entries, 0 to 327230
Series name: 운동량
Non-Null Count   Dtype 
--------------   ----- 
327231 non-null  object
dtypes: object(1)
memory usage: 2.5+ MB


In [65]:
df_copy.loc[2344] # 결측치가 아니라 '\N'이라는 object 데이터가 있음

대여일자      2020-01-20 00:00:00
대여소번호                     165
대여소명              165. 중앙근린공원
대여구분코드                 일일(회원)
성별                         \N
연령대코드                 AGE_003
이용건수                        1
운동량                        \N
탄소량                        \N
이동거리                      0.0
이용시간                       40
Name: 2344, dtype: object

In [66]:
df_copy['운동량'] = pd.to_numeric(df_copy['운동량'], errors='coerce')

In [67]:
df_copy['운동량'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 327231 entries, 0 to 327230
Series name: 운동량
Non-Null Count   Dtype  
--------------   -----  
326830 non-null  float64
dtypes: float64(1)
memory usage: 2.5 MB


### cut()

In [70]:
bins = [0, 100, 200, 300, df_copy['운동량'].max()]

In [71]:
bins

[0, 100, 200, 300, 163936052.3]

In [72]:
pd.cut(df_copy['운동량'], bins) # (컬럼, 나누는 구간) -> 어느 범주에 있는지

0                 (0.0, 100.0]
1                 (0.0, 100.0]
2         (300.0, 163936052.3]
3                 (0.0, 100.0]
4         (300.0, 163936052.3]
                  ...         
327226    (300.0, 163936052.3]
327227                     NaN
327228            (0.0, 100.0]
327229            (0.0, 100.0]
327230    (300.0, 163936052.3]
Name: 운동량, Length: 327231, dtype: category
Categories (4, interval[float64, right]): [(0.0, 100.0] < (100.0, 200.0] < (200.0, 300.0] < (300.0, 163936052.3]]

In [74]:
pd.cut(df_copy['운동량'], bins=10)
pd.cut(df_copy['운동량'], bins=10).value_counts()

운동량
(-163936.052, 16393605.23]      326816
(98361631.38, 114755236.61]          9
(32787210.46, 49180815.69]           2
(16393605.23, 32787210.46]           1
(114755236.61, 131148841.84]         1
(147542447.07, 163936052.3]          1
(81968026.15, 98361631.38]           0
(65574420.92, 81968026.15]           0
(49180815.69, 65574420.92]           0
(131148841.84, 147542447.07]         0
Name: count, dtype: int64

In [76]:
pd.qcut(df_copy['운동량'], q=10)
pd.qcut(df_copy['운동량'], q=10).value_counts()

운동량
(93.414, 192.02]           32690
(-0.001, 24.737]           32683
(24.737, 93.414]           32683
(601.705, 1079.744]        32683
(1889.606, 3328.186]       32683
(1079.744, 1889.606]       32683
(3328.186, 6805.188]       32683
(6805.188, 163936052.3]    32683
(344.45, 601.705]          32680
(192.02, 344.45]           32679
Name: count, dtype: int64

In [86]:
qcut_bins = [0, 0.2, 0.8, 1] # 직접 비율을 설정함

In [89]:
df_copy['운동량_한글'] = pd.qcut(df_copy['운동량'], qcut_bins, labels=['적음', '보통', '많음'])

In [90]:
df_copy.head()

대여일자  대여소번호      대여소명  대여구분코드  성별    연령대코드  이용건수      운동량    탄소량  \
0 2020-01-20      3      중랑센터  일일(회원)   M  AGE_003     3    61.82   0.52   
1 2020-01-20      3      중랑센터  일일(회원)   M  AGE_004     1    39.62   0.28   
2 2020-01-20      3      중랑센터      정기   M  AGE_005     3   430.85   4.01   
3 2020-01-20      5  상암센터 정비실  일일(회원)  \N  AGE_005     2     1.79   0.02   
4 2020-01-20      5  상암센터 정비실      정기   F  AGE_003     1  4501.96  45.47   

       이동거리  이용시간 운동량_한글  
0    2230.0    75     적음  
1    1220.0    15     적음  
2   17270.0    53     보통  
3      90.0    33     적음  
4  196010.0    64     많음